### PR TITLE
Enable cpp20 builds for DML EP and WinML API

### PR DESCRIPTION
**Pull request by smk2007**
Enable cpp20 builds for DML EP and WinML API

1) Missing typename for templated types
2) unmove helper for inline references to rvalue temporaries
This is okay since per the standard a temporary bound to a reference parameter in a function call exists until the end of the full expression containing that function call: if the function returns a reference, which outlives the full expression, it becomes a dangling reference.

3) static now not needed for template specializations

### CODE REVIEW

**Review by martinb35 — DISMISSED** [review]

**Comment by snnn** [comment]
Please fix the lint error.  You can download LLVM from https://github.com/llvm/llvm-project/releases then run the clang-format tool with the C/C++ source file, then it should be fine.

File: onnxruntime/core/providers/dml/DmlExecutionProvider/src/DmlCommon.h [hunk 1]
```diff
@@ -6,6 +6,11 @@
 #include <assert.h>
 #include "core/providers/dml/OperatorAuthorHelper/Common.h"
 
+template <class T>
+auto unmove_ptr(T&& t) {
+  return &static_cast<T&>(t);
```
Comment (fdwr) [inline onnxruntime/core/providers/dml/DmlExecutionProvider/src/DmlCommon.h:11]: [nit] Consistent indent with rest of file (no 🔎 needed to efficiently read the blocks of code :b).

File: winml/lib/Api.Image/inc/D3DDeviceCache.h [hunk 2]
```diff
@@ -104,9 +106,8 @@ class D3DDeviceCache {
   winrt::com_ptr<ID3D12RootSignature> tensorize_root_signature_;
   winrt::com_ptr<ID3D12RootSignature> detensorize_root_signature_;
 
-  winrt::com_ptr<ID3D12PipelineState>
-    cached_pipeline_state[PipelineStateCacheType::kCount][PipelineStateCacheFormat::kCount]
-                         [PipelineStateCacheFormat::kCount][PipelineStateCacheOperation::kCount];
+  winrt::com_ptr<ID3D12PipelineState> cached_pipeline_state[to_underlying_integer(PipelineStateCacheType::kCount)][to_underlying_integer(PipelineStateCacheFormat::kCount)][to_underlying_integer(PipelineStateCacheFormat::kCount)]
```
Comment (fdwr) [inline winml/lib/Api.Image/inc/D3DDeviceCache.h]: 🤔 Hmm, was more readable before spread across lines.

In cases where clangformat is being dumb, you can use `// clang-format off` (not that this necessarily needs it, but I've encountered that frequently with tables).

Comment (smk2007) [reply winml/lib/Api.Image/inc/D3DDeviceCache.h]: Yeah

**Review by fdwr — COMMENTED** [review]
🐿

**Pull request merged**